packages/gatsby/src/schema/extensions/index.js

```diff
@@ -0,0 +1,210 @@
+const {
+  GraphQLBoolean,
+  GraphQLNonNull,
+  GraphQLDirective,
+  GraphQLString,
+  DirectiveLocation,
+  defaultFieldResolver,
+} = require(`graphql`)
+const { GraphQLJSON } = require(`graphql-compose`)
+const report = require(`gatsby-cli/lib/reporter`)
+
+const { link, fileByPath } = require(`../resolvers`)
+const { getDateResolver } = require(`../types/date`)
+
+// Reserved for internal use
+const internalExtensionNames = [
+  `addDefaultResolvers`,
+  `createdFrom`,
+  `directives`,
+  `infer`,
+  `plugin`,
+]
+
+const typeExtensions = {
+  infer: {
+    description: `Infer field types from field values.`,
+    args: {
+      noDefaultResolvers: {
+        type: GraphQLBoolean,
+        description: `Don't add default resolvers to defined fields.`,
+        deprecationReason: `noDefaultResolvers is deprecated, annotate individual fields.`,
+      },
+    },
+  },
+  dontInfer: {
+    description: `Do not infer field types from field values.`,
+    args: {
+      noDefaultResolvers: {
+        type: GraphQLBoolean,
+        description: `Don't add default resolvers to defined fields.`,
+        deprecationReason: `noDefaultResolvers is deprecated, annotate individual fields.`,
+      },
+    },
+  },
+}
+
+const fieldExtensions = {
+  add: {
+    description: `Generic directive to add field extension.`,
+    args: {
+      extension: {
+        type: new GraphQLNonNull(GraphQLString),
+      },
+      options: {
+        type: GraphQLJSON,
+      },
+    },
+  },
+
+  addResolver: {
+    description: `Add a resolver specified by "type" to field.`,
+    args: {
+      type: {
+        type: new GraphQLNonNull(GraphQLString),
+        description:
+          `Type of the resolver. Types available by default are: ` +
+          `"dateformat", "link" and "fileByRelativePath".`,
+      },
+      options: {
+        type: GraphQLJSON,
+        description: `Resolver options. Vary based on resolver type.`,
+      },
+    },
+    process(args, fieldConfig) {
+      const { process } = fieldExtensions[args.type] || {}
+      if (typeof process === `function`) {
+        return process(args.options || {}, fieldConfig)
+      }
+      return {}
+    },
+  },
+
+  dateformat: {
```

Comment:
Not 100% sure we should add them separately.

Reply:
I don't think there are disadvantages to adding the shortcuts as well. The advantage otoh is that it allows a more concise way to add functionality to fields. E.g. adding foreign-key relations (which is probably the most common use-case) can be shortened to
```graphql
field: SomeType @link
```
instead of
```graphql
field: SomeType @addResolver(type: "link")
```

Reply:
I'm now considering not having addResolver, but having those instead.

The original motivation was that you'll have new resolvers to add, but API would be  the same with this extension api.